### PR TITLE
Add Record Create action in the frontend

### PR DESCRIPTION
**Pull request by Devessier**
In this PR:

- Updated the front-end types for workflows to include CRUD actions and global naming changes
- Allow users to create a Record Create action
- Scaffold the edit for Record Create action; for now, I render a `<VariableTagInput />` component for every editable field; it's likely we'll change it soon

Closes https://github.com/twentyhq/private-issues/issues/142

Demo:

https://github.com/user-attachments/assets/6f0b207a-b7d2-46d9-b5ab-9e32bde55d76

### CODE REVIEW

File: packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  const { activeObjectMetadataItems } = useFilteredObjectMetadataItems();
+
+  const availableMetadata: Array<SelectOption<string>> =
+    activeObjectMetadataItems.map((item) => ({
+      Icon: getIcon(item.icon),
+      label: item.labelPlural,
+      value: item.nameSingular,
+    }));
+
+  const form = useForm<SendEmailFormData>({
```
[the remark this answers is not in the document]
Comment (thomtrp) [reply packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx:44]: you have an example with WorkflowEditActionFormServerlessFunction

File: packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormServerlessFunction.tsx [hunk 2]
```diff
@@ -39,13 +39,13 @@ const StyledInputContainer = styled.div`
 
 type WorkflowEditActionFormServerlessFunctionProps =
   | {
-      action: WorkflowCodeStep;
+      action: WorkflowCodeAction;
```
Comment (thomtrp) [inline packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormServerlessFunction.tsx:42]: 💪

File: packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+          name={field.name}
+          control={form.control}
+          render={({ field: formField }) => (
+            <VariableTagInput
+              inputId={field.id}
+              label={field.label}
+              placeholder="Enter value (use {{variable}} for dynamic content)"
+              // @ts-expect-error Temporary fix
+              value={formField.value}
+              onChange={(value) => {
```
Comment (martmull) [inline packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx:154]: that need to be more complex. It works well for non composite fields, but it does not for composite fields like LINKS, ADDRESS, EMAILS, PHONES, ...
For those, you need a recursive Input function rendering, please, check what have been done with `renderFields` function in `packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormServerlessFunction.tsx`
See figma for the design -> https://www.figma.com/design/xt8O9mFeLl46C5InWwoMrN/Twenty?node-id=45225-479230&node-type=frame&t=spN9UE00UdvOjP4X-0
image

File: packages/twenty-front/src/modules/workflow/components/WorkflowStepDetail.tsx [hunk 4]
```diff
@@ -94,6 +96,19 @@ export const WorkflowStepDetail = ({
             />
           );
         }
+        case 'RECORD_CRUD': {
+          if (isWorkflowRecordCreateAction(stepDefinition.definition)) {
+            return (
+              <WorkflowEditActionFormRecordCreate
+                action={stepDefinition.definition}
+                // eslint-disable-next-line react/jsx-props-no-spreading
+                {...props}
```
Comment (martmull) [inline packages/twenty-front/src/modules/workflow/components/WorkflowStepDetail.tsx:105]: think we should avoid props spreading. It lower the readability of the code
just pass all 
```suggestion
      stepId={stepId}
      workflowVersion={workflowVersion}
      readonly={readonly}
      onTriggerUpdate={onTriggerUpdate}
      onActionUpdate={onActionUpdate}
```

**Review by martmull — COMMENTED** [review]
Hey, nice job, I have left some comments

File: packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx (hunk 3 above)
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx:154]: Thanks for mentioning. @thomtrp told me to create a first basic version as the desired implementation for fields was under discussion. We'll see what has been decided and maybe improve the implementation in a second PR.

Comment (thomtrp) [reply packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx:154]: @martmull @Devessier let's discuss the strategy for frontend on Monday. We will have to create new components for each fields. Having a common base will help creating those components in parallel. So yes, I would prefer having a simple version as a start 👌

File: packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx:44]: As far as I understand, the `WorkflowEditActionFormServerlessFunction` component directly sends debounced updates to the backend.

The form I implemented is a bit more complex as I want to synchronously update the available fields based on the selected object type while still debouncing the server update. I wouldn't be able to fully rely on `props.action` for that. Instead, I will have to create a `useState` specifically for the object type.

I will use a `useState` but I think we'll quickly want to create our own tiny form library.

[the remark this answers is not in the document]
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx:44]: I spotted a bug in the `WorkflowEditActionFormServerlessFunction` component. I created an issue here: https://github.com/twentyhq/twenty/issues/8523.

We'll have to use a `useState` in this component as well.

[the remark this answers is not in the document]
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx:44]: I will do the change on Monday.

File: packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx (hunk 3 above)
Comment (martmull) [reply packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx:154]: let's talk about that on today yes
Ok for basic version, but i suggest we don't introduce duplication
Let's do one basic version, used everywhere, and upgrade it in the future

File: packages/twenty-front/src/modules/workflow/components/WorkflowStepDetail.tsx (hunk 4 above)
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowStepDetail.tsx:105]: Here, the `props` object is a type union. If I provide the properties of this object individually, I will loose the type union and, for instance, `props.readonly` will be a boolean.

CleanShot 2024-11-18 at 15 01 41@2x

Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowStepDetail.tsx:105]: This component serves as a dispatcher. I think it's better to have strong types than to provide props as usually recommended.

Another way to keep the benefits without disabling the ESLint rule would be to have a single prop containing `readonly`, `onTriggerUpdate`, `onActionUpdate`:

```tsx
<WorkflowStepDetail
  xxx={{ readonly: true }}
/>

<WorkflowStepDetail
  xxx={{ onActionUpdate: yyy, onTriggerUpdate: zzz }}
/>
```

File: packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowEditActionFormRecordCreate.tsx]: Done

File: packages/twenty-front/src/modules/workflow/components/FormFieldInput.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import VariableTagInput from '@/workflow/search-variables/components/VariableTagInput';
+
+type FormFieldInputProps = {
+  recordFieldInputdId: string;
+  label: string;
+  value: string;
```
Comment (thomtrp) [inline packages/twenty-front/src/modules/workflow/components/FormFieldInput.tsx:6]: will probably come from a context but this is ok for now

File: packages/twenty-front/src/modules/workflow/components/FormFieldInput.tsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  onChange: (value: string) => void;
+  isReadOnly?: boolean;
+};
+
+export const FormFieldInput = ({
```
Comment (thomtrp) [inline packages/twenty-front/src/modules/workflow/components/FormFieldInput.tsx:11]: That component is not workflow specific. Could you move it next to `FieldInput.tsx` 🙏

File: packages/twenty-front/src/modules/workflow/components/FormFieldInput.tsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import VariableTagInput from '@/workflow/search-variables/components/VariableTagInput';
```
Comment (thomtrp) [inline packages/twenty-front/src/modules/workflow/components/FormFieldInput.tsx:1]: we need that component and the variables handling to be outside of workflows. But not for this PR!

**Review by thomtrp — APPROVED** [review]
LGTM, just move the FormFieldInput to the other folder please ✅

File: packages/twenty-front/src/modules/workflow/components/WorkflowStepDetail.tsx (hunk 4 above)
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/WorkflowStepDetail.tsx]: Done 👌

**Review by martmull — APPROVED** [review]
LGTM

**Pull request merged**